README.md

```diff
@@ -889,33 +901,43 @@ class C {
   m = 123;
 }
 
-C.prototype[Symbol.metadata].m;
-// { meta1: 1, meta2: 2 }
+C.prototype[Symbol.metadata][VALIDATIONS].public.m; // [1, 2]
 ```
 
-In addition, subsequent definitions to the same key will result in an array of values instead of a single value.
+Finally, the metadata object itself inherits from the parent class' metadata object if it exists, the `public` object inherits from the parent's `public` object if it exists, and the `private` array is a duplicate of the parent's `private` array (if it exists) with child values appended to the end:
 
 ```js
-function validateString(value, context) {
-  context.defineMetadata("validations", (value) => typeof value === "string"));
-}
+const MY_META = Symbol();
 
-function validateMaxLength(length) {
-  return (value, context) => {
-    context.defineMetadata("validations", (value) => value.length < length);
-  }
+function myMeta(value, context) {
+  context.defineMetadata(MY_META, 'metadata');
 }
 
+@myMeta
 class C {
-  @validateString
-  @validateMaxLength(10)
-  foo = "hello!";
+  @myMeta a = 123;
+  @myMeta #b = 456;
 }
 
-C.prototype[Symbol.metadata].foo.validations.length;
-// 2
+class D extends C {
+  @myMeta c = 123;
+  @myMeta #d = 456;
+}
+
+D[Symbol.metadata].constructor; // 'metadata'
+D.prototype[Symbol.metadata];
+// {
+//   public: {
+//     a: 'metadata',
+//     c: 'metadata',
+//   }
+
+//   private: ['metadata', 'metadata'],
+// }
 ```
 
+This inheritance is prototype based in the case of public fields and `constructor`, allowing users to distinguish between _own_ metadata and inherited metadata in these cases.
+
 This API design meets the following goals:
```

Comment:
Regarding some of the goals (listed below)

> - Metadata is easy to access, and it's possible to tell which class element it was associated with.

This falls apart a little with private members now given the names are no longer exposed in the metadata.  However, I think this is saved by the fact that the decorators themselves are now being provided with the name, and they could then potentially pass that down through the metadata value.

> - Metadata access is uniform, all metadata is accessed the same way. There is no need to learn a different technique for each type of class element.

I would say there are problems with this one because:

- Public metadata values can be inconsistent, are either single values or arrays of values.
- Private metadata values are accessed by array rather than a name-based mapping like public members
- While all private metadata is accessible via array no matter their source, I believe you would still have to walk the proto chain to get to all the public metadata for members that were overridden.

Reply:
- Public metadata values being inconsistent is something that users can design around, assuming metadata APIs are well formed in general. The alternative is to either create an array for each item, or to only allow one value and "last write wins". The former would be memory intensive, but is a possibility. The later is not possible as it makes certain use cases impractical. For the first draft of this proposal, we're going to stick with the design choice to avoid extra arrays/allocations, because we're trying to create the most performant spec possible.

-  Private metadata values actually _must_ be accessed differently. I realized this when creating the inheritance structure. Consider:

```js
class Foo {
  #x = 123;
}

class Bar extends Foo {
  #x = 456;
}
```

Unlike with public fields, where the subclass field would clobber/overwrite the superclass field, we actually have two separate fields here. Any decorations applied to each field would need to be applied independently. So, if we relied on mapping from the "name" of these fields, which is the same, and that map inherited like with public fields, users would likely correctly apply metadata to just the subclass field, without realizing they need to also apply it to the parent class field. That's why we decided to make private field metadata an array, ultimately, to prevent that issue in general.

- This is true, and part of the design. Public fields _can_ be overridden, as mentioned above, and decorators may-or-may-not apply any longer after being overridden. So, it's important to have the distinction. If decorations should still be applied to overridden fields, then users can manually crawl the prototype chain.

Reply:
> Unlike with public fields, where the subclass field would clobber/overwrite the superclass field, we actually have two separate fields here. Any decorations applied to each field would need to be applied independently. So, if we relied on mapping from the "name" of these fields, which is the same, and that map inherited like with public fields, users would likely correctly apply metadata to just the subclass field, without realizing they need to also apply it to the parent class field. That's why we decided to make private field metadata an array, ultimately, to prevent that issue in general.

Alternatively we could have it so that only metadata that is actually part of the class is included in private metadata, e.g. for your example, if we got the metadata on `Bar` only the private field inside `Bar` would be included. If people wanted all metadata of inherited fields they would simply walk the chain (e.g. `Object.getMetadata(Object.getPrototypeOf(Bar), myKey)`).

Reply:
Since the return override trick means you van install private fields on arbitrary objects, there’s no guarantee that prototype chain is available.